File: src/main/java/com/wesker926/leetcode/algorithms/p0496/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0496.s1;

import java.util.Arrays;
import java.util.LinkedList;

/**
 * @author wesker.gh
 * @date 2021/10/26
 * @description map + stack
 */
public class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] map = new int[10001];
        Arrays.fill(map, -1);
        LinkedList<Integer> stack = new LinkedList<>();
        for (int cur : nums2) {
            while (!stack.isEmpty() && stack.peekFirst() < cur) {
                map[stack.pop()] = cur;
            }
            stack.push(cur);
        }

        int[] result = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            result[i] = map[nums1[i]];
        }
        return result;
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0496/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/10/26
// @description: map + stack
package s1

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	var m [10001]int
	for i := 0; i < len(m); i++ {
		m[i] = -1
	}
	stack := make([]int, 0)

	for _, cur := range nums2 {
		for len(stack) > 0 && stack[len(stack)-1] < cur {
			m[stack[len(stack)-1]] = cur
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, cur)
	}

	result := make([]int, len(nums1))
	for i := 0; i < len(nums1); i++ {
		result[i] = m[nums1[i]]
	}
	return result
}
